taskfiles/utils.yml

```diff
@@ -141,3 +154,158 @@ tasks:
+    requires:
+      vars: ["BUILD_DIR", "INSTALL_PREFIX"]
+    cmds:
+      - >-
+        cmake
+        --install "{{.BUILD_DIR}}"
+        --prefix "{{.INSTALL_PREFIX}}"
+
+  # ===
+  # REMOTE UTILS
+  # ===
+
+  # Runs curl to download a file from the given URL.
+  #
+  # @param {string} URL
+  # @param {string} FILE_SHA256 Content hash to verify the downloaded file against.
+  # @param {string={{(base .URL)}}} [OUTPUT_FILE] Path where the file should be stored.
+  curl:
+    internal: true
+    label: "{{.TASK}}-{{.OUTPUT_FILE}}"
+    vars:
+      OUTPUT_FILE: "{{default (base .URL) .OUTPUT_FILE}}"
+    requires:
+      vars: ["FILE_SHA256", "URL"]
+    generates: ["{{.OUTPUT_FILE}}"]
+    status:
+      - >-
+        diff
+        <(echo "{{.FILE_SHA256}}")
+        <(openssl dgst -sha256 "{{.OUTPUT_FILE}}"
+        | awk '{print $2}')
+    cmds:
+      - |-
+        mkdir -p "{{dir .OUTPUT_FILE}}"
+        max_attempts=3
+        attempt=1
+        while [ $attempt -le $max_attempts ]; do
+          if curl \
+              --fail \
+              --location \
+              --show-error \
+              --connect-timeout 10 \
+              --max-time 300 \
+              "{{.URL}}" \
+              --output "{{.OUTPUT_FILE}}";
+          then
+            break
+          fi
+          attempt=$((attempt + 1))
+          sleep 5
+        done
+        if [ $attempt -gt $max_attempts ]; then
+          echo "Failed to download after $max_attempts attempts."
+          exit 1
+        fi
+
+  # Runs curl to download the tar file from the given URL and extracts its contents.
+  #
+  # @param {string} OUTPUT_DIR Directory in which to extract the tar file.
+  # @param {string} URL
+  # @param {string} FILE_SHA256 Content hash to verify the downloaded tar file against.
+  # @param {string={{.OUTPUT_DIR}}.md5} [CHECKSUM_FILE] File path to store the checksum of
+  # downloaded tar file.
+  # @param {string[]=[]} [EXCLUDE_PATTERNS] Path wildcard patterns that should not be extracted.
+  # @param {string[]=[]} [DATA_PATTERNS] Path wildcard patterns to extract.
+  # @param {int=1} [NUM_COMPONENTS_TO_STRIP] Number of leading path components to strip from the
+  # extracted files.
+  # @param {string={{.OUTPUT_DIR}}.tar.gz} [TAR_FILE] Path where the tar file should be stored.
+  download-and-extract-tar:
+    internal: true
+    label: "{{.TASK}}-{{.OUTPUT_DIR}}"
+    vars:
+      CHECKSUM_FILE: >-
+        {{default (printf "%s.md5" .OUTPUT_DIR) .CHECKSUM_FILE}}
+      DATA_PATTERNS:
+        ref: "default (list) .DATA_PATTERNS"
+      EXCLUDE_PATTERNS:
+        ref: "default (list) .EXCLUDE_PATTERNS"
+      NUM_COMPONENTS_TO_STRIP: "{{default 1 .NUM_COMPONENTS_TO_STRIP}}"
+      TAR_FILE: >-
+        {{default (printf "%s.tar.gz" .OUTPUT_DIR) .TAR_FILE}}
+    requires:
+      vars: ["FILE_SHA256", "OUTPUT_DIR", "URL"]
+    sources: ["{{.TASKFILE}}"]
+    generates: ["{{.CHECKSUM_FILE}}", "{{.TAR_FILE}}"]
```

Comment:
I need to double-check if these will cause issues.

Reply:
* I guess you want `.TASKFILE` to be a source so that if the way of extracting the file changes, it will be extracted again, right?
* I think `.TARFILE` should be a source (rather than a generated file) since it's the generated file of `curl`.